test/promise.cpp

```diff
@@ -10,6 +10,123 @@
 
 namespace {
 
+template<size_t Alignment, size_t Size = 4096>
+class static_arena {
+  std::array<uint8_t, Size> data_;
+  std::size_t offset_;
+
+  std::size_t offset_align() const {
+    std::size_t align = offset_ % Alignment;
+    if (align)
+	  return Alignment - align;
+
+    return 0;
+  }
+public:
+  static_arena() :
+      offset_(0)
+  { }
+
+  void* allocate(std::size_t bytes) {
+    auto align = offset_align();
+    bytes += align;
+    if ((offset_ + bytes) > data_.size())
+      throw std::bad_alloc();
+
+    void* result = reinterpret_cast<void*>(&data_[offset_+align]);
```

Comment:
Looks like most of the computation to satisfy alignment requirements can be done with `std::align` function (it's not available for gcc 4.9 so I have `pc::detail::align` which is just an alias to `std::align` for compilers which provide it and custom implementation for others):

    void* result = std::align(bytes, Alignment, &data_[offset_], data_.size() - offset_);
    if (!result)
        throw std::bad_alloc{};
    offset_ = reinterpret_cast<uint8_t*>(result) - data_.data();
    return result;